File: python/monitor_weights.py
#!/usr/bin/env python3
"""
Monitor survival and weight tracking.

Generates:
1. Kaplan-Meier survival curve with confidence intervals
2. Percent weight over time and raw weight over time
3. Spleen weight comparison if provided
"""

import os
import sys
import argparse
import logging
from datetime import datetime

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from lifelines import KaplanMeierFitter

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from functions.preprocessing import create_survival_table
from tools.plotting import plot_scatter, plot_multiple_comparisons, save_fig
from tools.file_io import read_excel_or_csv
from tools.df_tools import append_dataframe, fillna, rename_columns


def main():
    parser = argparse.ArgumentParser(description='Monitor weights and survival')
    parser.add_argument('-i', '--input-file', default='data/weights.xlsx',
                        help='Input weights file')
    parser.add_argument('-o', '--output-dir', default='figures/survival',
                        help='Output directory')
    parser.add_argument('-g', '--group-by', default='treatment,genotype',
                        help='Grouping columns')
    parser.add_argument('-w', '--week', action='store_true',
                        help='Use weeks instead of days')
    parser.add_argument('-p', '--optional-plots', action='store_true',
                        help='Generate optional plots')
    parser.add_argument('-t', '--troubleshooting', action='store_true')

    args = parser.parse_args()

    # Setup
    wd = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    metadata_cols = args.group_by.split(',')

    # Logging
    start_time = datetime.now()
    logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(message)s')
    logging.info(f'Script started at: {start_time}')

    # Read file
    logging.info('Reading data...')
    input_file = os.path.join(wd, args.input_file)
    weight_tbl = read_excel_or_csv(input_file)

    ext = os.path.splitext(input_file)[1].lower()
    if ext == '.xlsx':
        date_pattern = r'^\d+$'
    else:
        date_pattern = r'^\d{1,2}/\d{1,2}/\d{2,4}$'

    # Identify date columns
    is_date = weight_tbl.columns.str.match(date_pattern)
    id_cols = weight_tbl.columns[~is_date].tolist()
    date_cols = weight_tbl.columns[is_date].tolist()

    weight_tbl = weight_tbl[~weight_tbl['mouse_id'].isna()]

    # Derive group column
    metadata_cols = [c for c in metadata_cols if c in weight_tbl.columns]
    if len(metadata_cols) > 1:
        weight_tbl['group'] = weight_tbl[metadata_cols].apply(
            lambda x: ', '.join(x.astype(str)), axis=1
        )
    else:
        weight_tbl['group'] = weight_tbl[metadata_cols[0]]

    # Pivot to long format
    df = weight_tbl.melt(
        id_vars=[c for c in weight_tbl.columns if c not in date_cols],
        value_vars=date_cols,
        var_name='date',
        value_name='weight'
    )

    # Fix dates for Excel
    if ext == '.xlsx':
        df['date'] = pd.to_datetime(df['date'].astype(float), unit='D',
                                     origin='1899-12-30')
    else:
        df['date'] = pd.to_datetime(df['date'])

    # Plot survival
    logging.info('Plotting survival...')

    # Extract start and end info
    start_info = df.groupby('mouse_id').apply(
        lambda x: x.loc[x['date'].idxmin()]
    ).reset_index(drop=True)
    start_info = start_info[['mouse_id', 'group', 'date', 'weight']].copy()
    if 'spleen_weight' in df.columns:
        start_info = start_info.merge(
            weight_tbl[['mouse_id', 'spleen_weight']], on='mouse_id', how='left'
        )

    end_info = df[~df['weight'].isna()].groupby('mouse_id').apply(
        lambda x: x.loc[x['date'].idxmax()]
    ).reset_index(drop=True)
    end_info = end_info[['mouse_id', 'group', 'date', 'weight']].copy()

    survival = create_survival_table(start_info, end_info, end_date=df['date'].max())

    # Kaplan-Meier plot
    fig, ax = plt.subplots(figsize=(10, 6))

    time_col = 'week' if args.week else 'day'
    for group in survival['group'].unique():
        group_data = survival[survival['group'] == group]

        kmf = KaplanMeierFitter()
        kmf.fit(group_data[time_col], group_data['is_dead'], label=group)
        kmf.plot_survival_function(ax=ax, ci_show=True)

    ax.set_xlabel(f'Time ({"weeks" if args.week else "days"})')
    ax.set_ylabel('Survival Probability')
    ax.set_title('Survival Curve')
    ax.set_xlim(0, None)
    ax.set_ylim(0, 1)
    ax.legend(title='Group')

    if not args.troubleshooting:
        dirpath = os.path.join(wd, args.output_dir)
        os.makedirs(dirpath, exist_ok=True)
        fig.savefig(os.path.join(dirpath, 'survival-curve.png'),
                    dpi=300, bbox_inches='tight')
        plt.close(fig)

    # Plot weights
    logging.info('Plotting weights...')

    if not args.troubleshooting:
        dirpath = os.path.join(wd, args.output_dir, 'mouse-weights')
        os.makedirs(dirpath, exist_ok=True)

    # Merge start info for calculations
    df = df.merge(
        start_info[['mouse_id', 'date', 'weight']].rename(
            columns={'date': 'date_start', 'weight': 'weight_start'}
        ),
        on='mouse_id', how='left'
    )

    df['day'] = (df['date'] - df['date_start']).dt.days
    df['week'] = df['day'] / 7
    df['pct_weight'] = df['weight'] / df['weight_start']

    # Percent weight plot
    color_map = {}
    for group in df['group'].unique():
        if group in ['WT', 'H']:
            color_map[group] = '#FF7F0E'
        else:
            color_map[group] = 'rgba(23,190,207,1)'

    fig = plot_scatter(
        df,
        x='week' if args.week else 'day',
        y='pct_weight', group_by='group',
        ymin=0.75,
        xlabel=f'Time ({"weeks" if args.week else "days"})',
        ylabel='Percent Weight',
        title='Percent Weight Over Time',
        mode='lines+markers',
        color_discrete_map=color_map
    )

    if not args.troubleshooting:
        save_fig(fig, height=400, width=750, dirpath=dirpath,
                 filename='pct_weight', save_html=True)

    # Raw weight plot
    fig = plot_scatter(
        df,
        x='week' if args.week else 'day',
        y='weight', group_by='group',
        xlabel=f'Time ({"weeks" if args.week else "days"})',
        ylabel='Mouse Weight (g)',
        title='Raw Weight Over Time',
        mode='lines+markers',
        color_discrete_map=color_map
    )

    if not args.troubleshooting:
        save_fig(fig, height=400, width=600, dirpath=dirpath,
                 filename='raw_weight', save_html=True)

    # Weight loss at death
    survival['weight_ratio'] = survival['weight_end'] / survival['weight_start']

    try:
        fig = plot_multiple_comparisons(
            survival, x='group', y='weight_ratio',
            ymin=0.70,
            ylabel='[Final weight] / [Starting weight]',
            title='Weight Loss at Time of Death',
            show_numbers=True,
            test='fishers_lsd',
            custom_group_order=['WT', 'KO']
        )

        if not args.troubleshooting:
            fig.savefig(os.path.join(wd, args.output_dir, 'weight-loss.png'),
                        dpi=300, bbox_inches='tight')
            plt.close(fig)
    except Exception as e:
        logging.warning(f"Could not plot weight loss: {e}")

    # Spleen weight
    if 'spleen_weight' in weight_tbl.columns:
        logging.info('Plotting spleen weight...')

        if not args.troubleshooting:
            dirpath = os.path.join(wd, args.output_dir, 'spleen-weight')
            os.makedirs(dirpath, exist_ok=True)

        try:
            fig = plot_multiple_comparisons(
                survival, x='group', y='spleen_weight',
                ylabel='Spleen Weight (mg)',
                title='Spleen Weight',
                show_numbers=False,
                test='fishers_lsd',
                custom_group_order=['WT', 'KO']
            )

            if not args.troubleshooting:
                fig.savefig(os.path.join(dirpath, 'spleen_weight.svg'),
                            bbox_inches='tight')
                plt.close(fig)
        except Exception as e:
            logging.warning(f"Could not plot spleen weight: {e}")

    end_time = datetime.now()
    logging.info(f'Script ended at: {end_time}')
    logging.info(f'Script completed in: {end_time - start_time}')


if __name__ == '__main__':
    main()

File: R/monitor_weights.R
## Monitor survival
## 
## Facilitates survival monitoring by generating the following:
## 1. Kaplan-Meyer Survival curve with confidence intervals
## 2. Percent weight over time and raw weight over time
## 3. Spleen weight comparison if spleen_weight is provided in the weights table

wd = dirname(this.path::here())  # wd = '~/github/R/helperFlowCell'
library('optparse')
suppressPackageStartupMessages(library('logr'))
suppressPackageStartupMessages(library(dplyr))

import::from(magrittr, '%>%')
import::from(tibble, 'tibble')
import::from(tidyr, 'pivot_longer')
import::from(broom, 'tidy')
import::from(survival, 'survfit', 'Surv')
import::from(ggplot2,
    'ggplot', 'aes', 'geom_step', 'geom_ribbon',
    'labs', 'xlim', 'ylim', 'ggsave')
import::from(ggprism, 'theme_prism')

import::from(file.path(wd, 'R', 'functions', 'preprocessing.R'),
    'create_survival_table', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'plotting.R'),
    'plot_scatter', 'plot_multiple_comparisons', 'save_fig', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'file_io.R'),
    'read_excel_or_csv', .character_only=TRUE)
import::from(file.path(wd, 'R', 'tools', 'df_tools.R'),
    'append_dataframe', 'fillna', 'rename_columns', .character_only=TRUE)

# ----------------------------------------------------------------------
# Pre-script settings

# args
option_list = list(
    make_option(c("-i", "--input-file"), default='data/weights.xlsx',
                metavar='data/weights.xlsx', type="character",
                help="specify input file"),

    make_option(c("-o", "--output-dir"), default="figures/survival",
                metavar="figures/survival", type="character",
                help="set the output directory for the data"),

    make_option(c("-g", "--group-by"), default='treatment,genotype',
                metavar='treatment,genotype', type="character",
                help="enter a column or comma-separated list of columns, no spaces"),

    make_option(c("-w", "--week"), default=FALSE, action="store_true",
                metavar='FALSE', type="logical",
                help="use this to select week for the plots"),

    make_option(c("-p", "--optional-plots"), default=FALSE, action="store_true",
                metavar='FALSE', type="logical",
                help="use this to select optional plots"),

    make_option(c("-t", "--troubleshooting"), default=FALSE, action="store_true",
                metavar="FALSE", type="logical",
                help="enable if troubleshooting to prevent overwriting your files")
)
opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)
troubleshooting = opt[['troubleshooting']]

# args
metadata_cols <- unlist(strsplit(opt[['group-by']], ','))

# Start Log
start_time = Sys.time()
log <- log_open(paste0("monitor_weights-",
    strftime(start_time, format="%Y%m%d_%H%M%S"), '.log'))
log_print(paste('Script started at:', start_time))


# ----------------------------------------------------------------------
# Read file

log_print(paste(Sys.time(), 'Reading data...'))

input_file <- file.path(wd, opt[['input-file']])
weight_tbl <- read_excel_or_csv(input_file)

ext=tools::file_ext(input_file)
if (ext == 'xlsx') {
    date_pattern <- "^\\d+$"
} else if (ext == 'csv') {
    date_pattern <- "^\\d{1,2}/\\d{1,2}/\\d{2,4}$"
}
is_date <- grepl(date_pattern, colnames(weight_tbl))
id_cols <- colnames(weight_tbl)[!is_date]
date_cols <- colnames(weight_tbl)[is_date]
weight_tbl <- weight_tbl[(!is.na(weight_tbl[['mouse_id']])), ]

# Derive group col
metadata_cols <- intersect(metadata_cols, colnames(weight_tbl))
if (length(metadata_cols) > 1) {
    weight_tbl[['group']] <- apply( weight_tbl[ , metadata_cols ] , 1 , paste , collapse = ", " )
} else {
    weight_tbl[['group']] <- weight_tbl[[metadata_cols]]
}

# pivot
df = pivot_longer(
    data=weight_tbl,
    cols=all_of(date_cols),
    names_to='date',
    values_to='weight',
    values_drop_na = FALSE  # important
)

# fix dates
if (ext == 'xlsx') {
    df[['date']] <- as.Date(as.numeric(df[['date']]), origin = "1899-12-30")
}


# ----------------------------------------------------------------------
# Plot survival

log_print(paste(Sys.time(), 'Plotting survival...'))

# extract start and end info
start_info <- df %>%
  group_by(mouse_id) %>%
  slice_min(date, with_ties = FALSE) %>%
  ungroup() %>%
  select(any_of(c('mouse_id', 'spleen_weight', 'group', 'date', 'weight')))

end_info <- df %>%
    filter(!is.na(weight)) %>%
    group_by(mouse_id) %>%
    slice_max(order_by=date, n = 1, with_ties = FALSE) %>%
    ungroup() %>%
    select(all_of(c('mouse_id', 'group', 'date', 'weight')))

survival <- create_survival_table(start_info, end_info, end_date=max(df[['date']]))


# Get confidence intervals
if (opt[['week']]) {
    fit <- survfit(Surv(week, is_dead) ~ group, data = survival)
} else {
    fit <- survfit(Surv(day, is_dead) ~ group, data = survival)
}
surv_data <- tidy(fit)
surv_data[['strata']] <- gsub('group=', '', surv_data[['strata']])


# Plot
fig <- ggplot(surv_data, aes(x = time, y = estimate, color = strata)) +
    geom_step(linewidth = 1) +
    geom_ribbon(aes(ymin = conf.low, ymax = conf.high, fill = strata), alpha = 0.2, color = NA) +
    labs(x = paste("Time", if (opt[['week']]) {'(weeks)'} else {'(days)'}),
         y = "Survival Probability",
         color = "Genotype",
         fill = "Genotype",
         title = "Survival Curve") +
    xlim(0, NA) +
    ylim(0, 1) +
    theme_prism()

# save
if (!troubleshooting) {
    dirpath <- file.path(wd, opt[['output-dir']])
    if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
    ggsave(
        file.path(dirpath, 'survival-curve.png'),  
        plot=fig,
        height=4000, width=6000, dpi=500, units='px',
        scaling=2
    )
}


# ----------------------------------------------------------------------
# Plot Weights

log_print(paste(Sys.time(), 'Plotting weights...'))


# output_dir for this section
if (!troubleshooting) {
    dirpath <- file.path(wd, opt[['output-dir']], 'mouse-weights')
    if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
}


# Preprocessing
df <- merge(df,  start_info[, c('mouse_id', 'date', 'weight')],
    by="mouse_id", all.x=TRUE, all.y=TRUE, suffixes=c('', '_start'))

df[['day']] <- as.numeric(difftime(
    as.Date(df[['date']], format = '%m/%d/%y'),
    as.Date(df[['date_start']], format = '%m/%d/%y'),
    units = paste0('day', 's')
))
df[['week']] <- df[['day']]/7
df[['pct_weight']] <- df[['weight']] / df[['weight_start']]
df <- data.frame(df)
df <- append_dataframe(df,
    cbind(weight_tbl[, c('mouse_id', 'group')],
        data.frame(
            day=max(df[['day']]+1),
            week=max(df[['week']]+1/7)
        )
    )
)
df <- df[order(df[['mouse_id']], desc(df[['day']])), ]

# set colors
# TODO: un-hardcode this
color_map <- data.frame(unique(df[c('mouse_id', 'group')]))
color_map[['color']] <- lapply(
    color_map[['group']],
    function(x) if ((x=='WT') | (x=='H')) {
        '#FF7F0E'  # orange
    } else if (x=='M') {
        '#377EB8'  # blue
    } else if (x=='PBS') {
        '#9467bd'  # purple
    } else {
        'rgba(23,190,207,1)'  # sea green
    }
)
color_discrete_map <- setNames(color_map[['color']], color_map[['group']])


# Percent weight
fig <- plot_scatter(
    df,
    x=if (opt[['week']]) {'week'} else {'day'},
    y='pct_weight', group_by='group',
    ymin=0.75,
    xlabel=paste("Time", if (opt[['week']]) {'(weeks)'} else {'(days)'}),
    ylabel='Percent Weight',
    title='Percent Weight Over Time',
    mode='lines+markers',
    color_discrete_map=color_discrete_map,
    hover_data=c('mouse_id', 'sex', 'genotype', 'dob', 'day', 'week')
)

# save
if (!troubleshooting) {
    save_fig(
        fig=fig,
        height=400, width=750,
        dirpath=dirpath,
        filename='pct_weight',
        save_html=TRUE
    )
}

# Raw weight
fig <- plot_scatter(
    df,
    x=if (opt[['week']]) {'week'} else {'day'},
    y='weight', group_by='group',
    xlabel=paste("Time", if (opt[['week']]) {'(weeks)'} else {'(days)'}),
    ylabel='Mouse Weight (g)',
    title='Raw Weight Over Time',
    mode='lines+markers',
    color_discrete_map=color_discrete_map,
    hover_data=c('sex', 'genotype', 'dob', 'day', 'week')
)

# save
if (!troubleshooting) {
    save_fig(
        fig=fig,
        height=400, width=600,
        dirpath=dirpath,
        filename='raw_weight',
        save_html=TRUE
    )
}


# Weight loss at time of death
survival[['weight_ratio']] <- survival[['weight_end']] / survival[['weight_start']] 

fig <- plot_multiple_comparisons(
    survival,
    x='group', y='weight_ratio',
    ymin=0.70,
    ylabel="[Final weight (g)] / [Starting weight (g)]",
    title='Weight Loss at Time of Death',
    show_numbers=TRUE,
    test='fishers_lsd',
    custom_group_order=intersect(c('WT', 'KO'), unique(weight_tbl[['group']]))
)

# save
if (!troubleshooting) {
    withCallingHandlers({
        ggsave(
            file.path(wd, opt[['output-dir']], 'weight-loss.png'),
            plot=fig,
            height=2000, width=2400, dpi=300, units='px',
            scaling=1
        )
    }, warning = function(w) {
        if ( any(grepl("containing non-finite values", w),
                 grepl("outside the scale range", w),
                 grepl("fewer than two data points", w),
                 grepl("argument is not numeric or logical: returning NA", w)) ) {
            invokeRestart("muffleWarning")
        }
    })
}


# ----------------------------------------------------------------------
# Plot spleen weight


if ('spleen_weight' %in% colnames(weight_tbl)) {
    
    # output_dir for this section
    if (!troubleshooting) {
        dirpath <- file.path(wd, opt[['output-dir']], 'spleen-weight')
        if (!dir.exists(dirpath)) { dir.create(dirpath, recursive=TRUE) }
    }

    # ----------------------------------------------------------------------
    # Spleen Weight by Group

    fig <- plot_multiple_comparisons(
        survival,
        x='group', y='spleen_weight',
        ylabel="Spleen Weight (mg)",
        title='Spleen Weight',
        show_numbers=FALSE,
        test='fishers_lsd',
        custom_group_order=intersect(c('WT', 'KO'), unique(weight_tbl[['group']]))
    )

    # save
    if (!troubleshooting) {
        withCallingHandlers({
            ggsave(
                file.path(wd, opt[['output-dir']], 'spleen-weight', 'spleen_weight.svg'),
                plot=fig,
                height=5000, width=5000, dpi=500, units='px',
                scaling=1
            )
        }, warning = function(w) {
            if ( any(grepl("containing non-finite values", w),
                     grepl("outside the scale range", w),
                     grepl("fewer than two data points", w),
                     grepl("argument is not numeric or logical: returning NA", w)) ) {
                invokeRestart("muffleWarning")
            }
        })
    }


    # ----------------------------------------------------------------------
    # Spleen Weight Ratio vs. Treatment Duration

    fig <- plot_scatter(
        survival,
        x=if (opt[['week']]) {'week'} else {'day'},
        y='spleen_weight',
        group_by='group',
        ymin=0,
        xlabel=paste("Time of Death", if (opt[['week']]) {'(weeks)'} else {'(days)'}),
        ylabel='Spleen Weight (mg)',
        title='Spleen Weight vs. Treatment Duration',
        mode='markers'
    )

    # save
    if (!troubleshooting) {
        save_fig(
            fig=fig,
            height=350, width=500,
            dirpath=dirpath,
            filename='spleen_weight_vs_treatment_duration',
            save_html=TRUE
        )
    }


    # ----------------------------------------------------------------------
    # Spleen Weight vs. Mouse Starting Weight

    fig <- plot_scatter(
        survival,
        x='weight_start', y='spleen_weight', group_by='group',
        ymin=0,
        xlabel='Mouse Starting Weight (g)', ylabel='Spleen Weight (mg)',
        title='Spleen Weight vs. Mouse Starting Weight',
        mode='markers',
        # color_discrete_map=color_discrete_map,
        hover_data=c('sex', 'group')
    )

    # save
    if (!troubleshooting) {
        save_fig(
            fig=fig,
            height=350, width=500,
            dirpath=dirpath,
            filename='spleen_vs_mouse_weight',
            save_html=TRUE
        )
    }


    # optional plots
    if (opt[['optional-plots']]) {

        # ----------------------------------------------------------------------
        # Spleen Weight Ratio vs. Treatment Duration

        survival[['spleen_weight_ratio_start']] <- survival[['spleen_weight']] / survival[['weight_start']]

        fig <- plot_scatter(
            survival,
            x=if (opt[['week']]) {'week'} else {'day'},
            y='spleen_weight_ratio_start',
            group_by='group',
            ymin=0,
            xlabel=paste("Time of Death", if (opt[['week']]) {'(weeks)'} else {'(days)'}),
            ylabel='[Spleen Weight / Mouse Starting Weight] (mg/g)',
            title='Spleen Weight Ratio vs. Treatment Duration',
            mode='markers'
        )


        # save
        if (!troubleshooting) {
            save_fig(
                fig=fig,
                height=400, width=600,
                dirpath=dirpath,
                filename='spleen_weight_ratio_start_vs_treatment_duration',
                save_html=TRUE
            )
        }

        # ----------------------------------------------------------------------
        # Spleen Weight Ratio vs. Treatment Duration

        survival[['spleen_weight_ratio_end']] <- survival[['spleen_weight']] / survival[['weight_end']]

        fig <- plot_scatter(
            survival,
            x=if (opt[['week']]) {'week'} else {'day'},
            y='spleen_weight_ratio_end',
            group_by='group',
            ymin=0,
            xlabel=paste("Time of Death", if (opt[['week']]) {'(weeks)'} else {'(days)'}),
            ylabel='[Spleen Weight / Mouse Final Weight] (mg/g)',
            title='Spleen Weight Ratio vs. Treatment Duration',
            mode='markers'
        )

        # save
        if (!troubleshooting) {
            save_fig(
                fig=fig,
                height=400, width=600,
                dirpath=dirpath,
                filename='spleen_weight_ratio_final_vs_treatment_duration',
                save_html=TRUE
            )
        }
    }


}

end_time = Sys.time()
log_print(paste('Script ended at:', Sys.time()))
log_print(paste("Script completed in:", difftime(end_time, start_time)))
log_close()
